### DISCUSSION TITLE
Python starter sample code for downloading API data

### DISCUSSION THREAD

**Discussion by psatVol** [ercot/api-specs#34] [Q&A] [answered]
Hi,

I am trying to download data from: https://data.ercot.com/data-product-archive/NP4-732-CD

Does anyone have a sample starter code they don't mind sharing for this?

Thanks

**Comment by mackermann-ercot** [comment, marked as answer]
Hi there, 

After installing the _Requests_ package (via pip), you can use the following code:

```
import requests

# USER MUST PROVIDE THIS INFORMATION
#
USERNAME = "USERNAME"
PASSWORD = "PASSWORD"
SUBSCRIPTION_KEY = "SUBSCRIPTION_KEY"

# Your specific API endpoint (could also be passed as a command line argument)
#
ARCHIVE_URL = "https://api.ercot.com/api/public-reports/archive"
ARCHIVE_TARGET = "NP4-732-CD"

# Authorization URL for signing into ERCOT Public API account
#
AUTH_URL = "https://ercotb2c.b2clogin.com/ercotb2c.onmicrosoft.com/B2C_1_PUBAPI-ROPC-FLOW/oauth2/v2.0/token\
?username={username}\
&password={password}\
&grant_type=password\
&scope=openid+fec253ea-0d06-4272-a5e6-b478baeecd70+offline_access\
&client_id=fec253ea-0d06-4272-a5e6-b478baeecd70\
&response_type=id_token"

# Sign In/Authenticate
#
auth_response = requests.post(AUTH_URL.format(username = USERNAME, password=PASSWORD))

# Retrieve access token
#
access_token = auth_response.json().get("access_token")

# Use Subscription key and bearer token to retrieve all Public API Reports
#
headers = {"Authorization": "Bearer " + access_token, "Ocp-Apim-Subscription-Key": SUBSCRIPTION_KEY}
product_response = requests.get(ARCHIVE_URL + "/" + ARCHIVE_TARGET, headers=headers)

print(product_response.text)
```

To install the Requests package:
`python -m pip install Requests`

I hope this helps!  Enjoy.

Regards,
mackermann, ERCOT

**Reply by psatVol** [reply]
Thanks

**Reply by CaffeineLab** [reply]
Are you modifying that token request url at all?  You need the entire thing, leave the client_id and scope alone.  If you modify or omit the client_id value, then you will get the error:    

{"error":"invalid_client","error_description":"AADB2C90018: The client id 'WHATEVERYOUPUTINCLIENT_IDWILLBEHERE' specified in the request is not registered in tenant 'ercotb2c.onmicrosoft.com'.\r\nCorrelation ID: f6761cf9-e8d0-4c6e-88b7-d0d06d8c5c0e\r\nTimestamp: 2024-12-02 01:07:59Z\r\n"}

Also - do the very basic CURL request that I posted here.  If you get the two requests to work - then understanding why they don't work in postman/python/c/go/etc will become much easier....

[CURL Example](https://github.com/ercot/api-specs/discussions/66)

**Reply by DMJ-CV-91913** [reply]
I revised my code just to see if it would return a JSON response and nothing( i was able to retrieve the download links of archived reports, and download csv files, but not public reports and I would like to retrieve via json if possible.   i have defined my credentials in variables.  Do I need to build a parametrized query once an access token is returned to specify begin and end dates for json api ?  And it is written in PHP, using CURL.


// Authorization URL for signing into ERCOT Public API account
$auth_url = "https://ercotb2c.b2clogin.com/ercotb2c.onmicrosoft.com/B2C_1_PUBAPI-ROPC-FLOW/oauth2/v2.0/token";

// URL encode the username and password to ensure they are properly formatted
//$archive_url = "https://api.ercot.com/api/public-reports/archive";
//$archive_url = "https://api.ercot.com/api/public-reports/np4-732-cd/wpp_hrly_avrg_actl_fcast";
$archive_url = "https://api.ercot.com/api/public-reports";
$archive_target = "NP4-732-CD";
$report_txt = "wpp_hrly_avrg_actl_fcast";

// Authorization parameters
$auth_params = [
    'username' => $username,
    'password' => $password,
    'grant_type' => 'password',
    'scope' => "openid {$auth_orig} offline_access",
    'client_id' => $auth_orig,
    'response_type' => 'id_token'
];

// Build the query string
$query_string = http_build_query($auth_params);

// Initialize cURL for the authentication request
$ch1 = curl_init();
// Set cURL options

curl_setopt($ch1, CURLOPT_URL, $auth_url . '?' . $query_string);
//curl_setopt($ch1, CURLOPT_RETURNTRANSFER, 1);
//curl_setopt($ch1, CURLOPT_POST, 1);
curl_setopt($ch1, CURLOPT_POST, true); // Specify POST method
curl_setopt($ch1, CURLOPT_RETURNTRANSFER, true); // Return response

$response_1 = curl_exec($ch1);

// Check for cURL errors
if (curl_errno($ch1)) {
    echo 'cURL Error: ' . curl_error($ch1);
    exit;
}
curl_close($ch1);

// Decode the response
$response_data_1 = json_decode($response_1, true);

// Debugging output for authentication response
echo "Authentication Response: \n";
print_r($response_data_1);



// Check for an access token
if (!isset($response_data_1['access_token'])) {
    echo "Error: Unable to retrieve access token.\n";
    print_r($response_data_1); // Debugging output if the token isn't retrieved
    exit;
}

$access_token = $response_data_1['access_token'];
$id_token = $response_data_1['id_token'];

// Debugging output for access token

echo "Access Token: $access_token\n";

echo "ID Token: $id_token\n";

echo "subscription-key: $subscription_key\n";



// Prepare headers for the next API call
$headers = [
    "Authorization: Bearer " . $id_token,
    "Ocp-Apim-Subscription-Key: " . $subscription_key
];


// Request the public API data

// Authorization parameters
$qry_params = [
	'deliveryDateFrom' => '2024-01-01',
	'deliveryDateTo' => '2024-10-31'
];


$public_url = "{$archive_url}/{$archive_target}/{$report_txt}";
//https://api.ercot.com/api/public-reports/NP4-732-CD/wpp_hrly_avrg_actl_fcast
// Build the query string
$query_string_2 = http_build_query($qry_params);

$public_url_2 = "{$archive_url}/{$archive_target}/{$report_txt}?{$query_string_2}";

// Full URL with query parameters
$full_url = $public_url . '?' . $query_string_2;

// Print full URL for debugging
print_r($full_url);


$ch = curl_init();
curl_setopt($ch, CURLOPT_URL, $full_url);
curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
curl_setopt($ch, CURLOPT_FOLLOWLOCATION, true); // Follow redirects
curl_setopt($ch, CURLOPT_HTTPGET, true); // Explicitly specify GET method
curl_setopt($ch, CURLOPT_HTTPHEADER, $headers);

// Execute the API call
$product_response = curl_exec($ch);

// Check for cURL errors
if (curl_errno($ch)) {
    echo "cURL Error: " . curl_error($ch);
    curl_close($ch);
    exit;
}

// Check HTTP status code
$http_code = curl_getinfo($ch, CURLINFO_HTTP_CODE);
if ($http_code !== 200) {
    echo "Error: API returned HTTP status code $http_code.";
    curl_close($ch);
    exit;
}

// Close cURL
curl_close($ch);

// Check if the response is empty
if (empty($product_response)) {
    echo "Error: Empty response from the API.";
    exit;
}

// Debug raw response
echo "Raw Response: " . $product_response;

// Decode the JSON response
$data = json_decode(trim($product_response), true);

// Check for JSON decoding errors
if (json_last_error() !== JSON_ERROR_NONE) {
    echo "Error decoding JSON: " . json_last_error_msg();
    exit;
}

// Print the decoded data
print_r($data);

exit;

**Reply by ytangpe** [reply]
Hi  mackermann,
I ran the following code for about one month, 
Then I got error:
{'error': 'invalid_request', 'error_description': 'AADB2C90083: The request is missing required parameter: grant_type.\r\nCorrelation ID: e81714e3-3810-497f-b5cc-895f0b8702ee\r\nTimestamp: 2025-04-29 17:20:18Z\
r\n'}

I checked my account, it says. 
You have reached your maximum subscriptions. 

Do you have any idea?

Best Regards.



> Hi there,
> 
> After installing the _Requests_ package (via pip), you can use the following code:
> 
> ```
> import requests
> 
> # USER MUST PROVIDE THIS INFORMATION
> #
> USERNAME = "USERNAME"
> PASSWORD = "PASSWORD"
> SUBSCRIPTION_KEY = "SUBSCRIPTION_KEY"
> 
> # Your specific API endpoint (could also be passed as a command line argument)
> #
> ARCHIVE_URL = "https://api.ercot.com/api/public-reports/archive"
> ARCHIVE_TARGET = "NP4-732-CD"
> 
> # Authorization URL for signing into ERCOT Public API account
> #
> AUTH_URL = "https://ercotb2c.b2clogin.com/ercotb2c.onmicrosoft.com/B2C_1_PUBAPI-ROPC-FLOW/oauth2/v2.0/token\
> ?username={username}\
> &password={password}\
> &grant_type=password\
> &scope=openid+fec253ea-0d06-4272-a5e6-b478baeecd70+offline_access\
> &client_id=fec253ea-0d06-4272-a5e6-b478baeecd70\
> &response_type=id_token"
> 
> # Sign In/Authenticate
> #
> auth_response = requests.post(AUTH_URL.format(username = USERNAME, password=PASSWORD))
> 
> # Retrieve access token
> #
> access_token = auth_response.json().get("access_token")
> 
> # Use Subscription key and bearer token to retrieve all Public API Reports
> #
> headers = {"Authorization": "Bearer " + access_token, "Ocp-Apim-Subscription-Key": SUBSCRIPTION_KEY}
> product_response = requests.get(ARCHIVE_URL + "/" + ARCHIVE_TARGET, headers=headers)
> 
> print(product_response.text)
> ```
> 
> To install the Requests package: `python -m pip install Requests`
> 
> I hope this helps! Enjoy.
> 
> Regards, mackermann, ERCOT

**Reply by ytangpe** [reply]
I use ChatGpt to update above code, then It works. looks Ercot update their system.

**Reply by MJuddBooth** [reply]
I will just record for posterity that microsoft does not like the plusses in the `scope` parameter to be url-encoded, so one cannot just pass a dictionary as the `data` keyword to requests, but must instead use `urllib.parse.encode(params, safe="+")` to generate a string to pass as the `data` argument to `requests.post`

**Reply by CaffeineLab** [reply]
I've also been meaning to write in here that you shouldn't pass the username and password as part of the querystring at all.  You should put it in the POST data...   I know I copied the same example code and have been doing this for quite some time - I need to fix the example I posted here as well....  

Just remove the username and password from the AUTH_URL and then move it into the post data

```
            auth_response = requests.post(
                url=AUTH_URL ,
                data={
                    'username': username,
                    'password': password,
                },
                timeout=15
            )
```

**Comment by aes18** [comment]
I am trying to extract data from https://api.ercot.com/api/public-reports/archive/NP4-732-CD  for specific dates, say from June 1, 2024 to June 30 2024, and store it all in one dataframe. The API response provides metadata, but when I attempt to download data from the links in this response, I encounter the following error:

Failed to download from https://api.ercot.com/api/public-reports/archive/NP4-732-CD?download=1019946422: 400 Client Error: Bad Request for URL: https://api.ercot.com/api/public-reports/archive/NP4-732-CD?download=1019946422 Status Code: 400 Response Content: {"timestamp":"2024-06-15 05:57:48","code":400,"status":"BAD_REQUEST","message":"Your request could not be processed. The request may be malformed or the URL does not exist."}    

Can someone help me with this?

**Reply by mackermann-ercot** [reply]
Hello aes18,

Thanks for the question.  I have duplicated your request several times and did not receive the same error.  The file downloaded as expected, but it's possible that you may have network latency in your request (i.e. timed out) or the request may be blocked by your local firewall.  Are you currently having to use any proxy settings to connect to external sites?

Regards,
-mackermann, ERCOT

**Comment by aes18** [comment]
Hello Mackermann,

Thanks for your response .No, I am not using any proxy settings. I am running this script in an Azure workspace and have tried several times, but I keep encountering the same error. I'm not sure if my approach is correct, so please correct me or guide me if there is a better way to do this. Here is the code I used:

headers = {"Authorization": "Bearer " + access_token, "Ocp-Apim-Subscription-Key": SUBSCRIPTION_KEY }
download_links = [ archive['_links']['endpoint']['href']  for archive in response_data['archives'] ]

def download_and_extract_csv(url, headers):
    try:
        response = requests.get(url, headers=headers)
        response.raise_for_status()  
        content_type = response.headers.get('Content-Type')
  
        if 'application/zip' in content_type:
            with ZipFile(BytesIO(response.content)) as zip_file:
           
                csv_files = [zip_file.open(name) for name in zip_file.namelist() if name.endswith('.csv')]
                dataframes = [pd.read_csv(csv_file) for csv_file in csv_files]
                return pd.concat(dataframes, ignore_index=True) if dataframes else None
        
        print(f"Unexpected content type {content_type} for URL: {url}")
        print(f"Response Content: {response.content.decode('utf-8')}")
        return None
    except requests.exceptions.RequestException as e:
        print(f"Failed to download from {url}: {e}")
        print(f"Status Code: {response.status_code}")
        print(f"Response Content: {response.content.decode('utf-8')}")
        return None
    except pd.errors.ParserError as e:
        print(f"Failed to parse CSV from {url}: {e}")
        print(f"Response Content: {response.content.decode('utf-8')}")
        return None

dataframes = [download_and_extract_csv(link, headers) for link in download_links if download_and_extract_csv(link, headers) is not None]
if dataframes:
    combined_df = pd.concat(dataframes, ignore_index=True)

**Comment by mackermann-ercot** [comment]
Hello aes18,

We have recently experienced some performance issues with the overall system for the past several days and have since employed solutions to mitigate the issues (earlier today 7/16).  You could have been experiencing the 400 errors as a timeout during that period.  Would you please run your code again and see if you are still seeing the issue?

Regards,
-mackermann, ERCOT

**Comment by venkatsu9** [comment]
I have a similar error message.  When I run the python code in the June 19 example given above after substituting the username, password and subscription key with my own, I get a list of metadata.  However, when I click on any of the links in the metadata, I get the following error "{ "statusCode": 401, "message": "Access denied due to missing subscription key. Make sure to include subscription key when making requests to an API."

**Reply by CaffeineLab** [reply]
The links aren't "clickable".  You need to supply the bearer token with the request for the document.

`headers = {"Authorization": "Bearer " + access_token, "Ocp-Apim-Subscription-Key": SUBSCRIPTION_KEY}`
